devices/flash-flexspi/flashdrv.c

```diff
@@ -0,0 +1,381 @@
+
+	if (dev == NULL || dev->active == 0) {
+		return -ENXIO;
+	}
+
+	/* Initial sector value check, nothing to do */
+	if (dev->sectorPrevAddr == (addr_t)-1) {
+		return EOK;
+	}
+
+	if (dev->sectorPrevAddr == dev->sectorSyncAddr) {
+		return EOK;
+	}
+
+	/* all 'ones' in buffer means sector is erased ... */
+	for (pos = 0; pos < dev->nor->pageSz; ++pos) {
+		if (*(dev->sectorBuf + pos) != NOR_ERASED_STATE) {
+			break;
+		}
+	}
+
+	/* ... than erase may be skipped */
+	if (pos != dev->nor->sectorSz) {
+		res = nor_eraseSector(&dev->fspi, dev->port, sectorAddr, dev->timeout);
+		if (res < 0) {
+			return res;
+		}
+	}
+
+
+	for (ofs = 0; ofs < dev->nor->sectorSz; ofs += dev->nor->pageSz) {
+		/* Just erased NOR page contains all 'ones' */
+		for (pos = 0; pos < dev->nor->pageSz; ++pos) {
+			if (*(dev->sectorBuf + ofs + pos) != NOR_ERASED_STATE) {
+				break;
+			}
+		}
+
+		/* then if buffer is the same skip single page program */
+		if (pos == dev->nor->pageSz) {
+			continue;
+		}
+
+		res = nor_pageProgram(&dev->fspi, dev->port, sectorAddr + ofs, dev->sectorBuf + ofs, dev->nor->pageSz, dev->timeout);
+		if (res < 0) {
+			return res;
+		}
+	}
+
+	hal_cpuInvCache(hal_cpuDCache, sectorAddr, dev->nor->sectorSz);
+
+	dev->sectorSyncAddr = dev->sectorPrevAddr;
+
+	return dev->nor->sectorSz;
+}
+
+
+static ssize_t bufferSync(unsigned int minor, addr_t dstAddr)
+{
+	ssize_t res = 0;
+	struct nor_device *dev = minorToDevice(minor);
+	const int sectorLast = get_sectorIdFromAddress(dev, dev->sectorPrevAddr);
+	const int sectorCurr = get_sectorIdFromAddress(dev, dstAddr);
+
+	if (sectorCurr != sectorLast) {
+		res = flashdrv_sync(minor);
+		if (res < 0) {
+			return res;
+		}
+
+		res = nor_readData(&dev->fspi, dev->port, get_sectorAddress(dev, dstAddr), dev->sectorBuf, dev->nor->sectorSz, dev->timeout);
+		if (res < 0) {
+			return res;
+		}
+
+		dev->sectorPrevAddr = dstAddr;
+	}
+
+	return res;
+}
+
+
+static ssize_t flashdrv_write(unsigned int minor, addr_t dstAddr, const void *data, size_t size)
+{
+	int res;
+	addr_t ofs;
+	const u8 *srcPtr = data;
+	size_t chunkSz, doneBytes;
+	struct nor_device *dev = minorToDevice(minor);
+
+	if (dev == NULL || dev->active == 0)
+		return -ENXIO;
+
+	if (data == NULL || dstAddr + size > dev->fspi.slFlashSz[dev->port])
+		return -EINVAL;
+
+	if (size == 0)
+		return 0;
+
+	doneBytes = 0;
+	while (doneBytes < size) {
+		res = bufferSync(minor, dstAddr);
+		if (res < 0) {
+			return res; /* FIXME: should error or doneBytes to be returned ? */
+		}
+
+		chunkSz = size - doneBytes;
+		ofs = dstAddr & (dev->nor->sectorSz - 1);
+
+		if (chunkSz > dev->nor->sectorSz - ofs) {
+			chunkSz = dev->nor->sectorSz - ofs;
+		}
+
+		hal_memcpy(dev->sectorBuf + ofs, srcPtr, chunkSz);
+
+		dstAddr += chunkSz;
+		srcPtr += chunkSz;
+		doneBytes += chunkSz;
+	}
+
+	if (doneBytes > 0) {
+		res = bufferSync(minor, dstAddr);
+		if (res < 0) {
+			return res; /* FIXME: should error or doneBytes to be returned ? */
```

Comment:
Perhaps done bytes, so we can try again writing rest?

Reply:
If buffer sync failed that is not a good sign, that's why I'm torn here `?`.

Reply:
Ok, I think I agree that error should be returned - write() is behaving this way (in case of IO error).